### PR TITLE
Add tool-centric API endpoint to install dependencies

### PR DESCRIPTION
**Pull request by mvdbeek**
This was added 16.10

### CODE REVIEW

**Comment by afgane** [comment]
Can (or should) a test be written for this?
You can isolate it so it would run only on Galaxy 16.10 with this decorator `@test_util.skip_unless_galaxy('release_16.10')`

File: bioblend/galaxy/tools/__init__.py [hunk 1]
```diff
@@ -66,6 +66,19 @@ def _raw_get_tool(self, in_panel=None, trackster=None):
         params['trackster'] = trackster
         return self._get(params=params)
 
+    def install_dependencies(self, tool_id):
+        """
+        Install dependencies for a given tool via a resolver.
+        This works for conda only currently.
+        This functionality is available since galaxy release 16.10
+        and is limited to admins.
```
Comment (nsoranzo) [inline bioblend/galaxy/tools/__init__.py]: `and is available only to Galaxy admins.`

File: bioblend/galaxy/tools/__init__.py [hunk 2]
```diff
@@ -66,6 +66,19 @@ def _raw_get_tool(self, in_panel=None, trackster=None):
...
+
+        :type tool_id: str
+        :param tool_id: id of the requested tool
+        """
+        url = "%s/tools/%s/install_dependencies" % (self.gi.url, tool_id)
+        return Client._post(self, url=url, payload={})
```
Comment (nsoranzo) [inline bioblend/galaxy/tools/__init__.py]:
```python
return self._post(payload={}, url=url)
```

**Comment by mvdbeek** [comment]
> Can (or should) a test be written for this?
> You can isolate it so it would run only on Galaxy 16.10 with this decorator @test_util.skip_unless_galaxy('release_16.10')

The API itself is being tested in test/integration/test_resolvers.py, which sets up a conda environment. Can we do this in bioblend?

File: bioblend/galaxy/tools/__init__.py [hunk 3]
```diff
@@ -66,6 +66,19 @@ def _raw_get_tool(self, in_panel=None, trackster=None):
...
+        Install dependencies for a given tool via a resolver.
+        This works only for Conda currently.
+        This functionality is available since Galaxy release_16.10
+        and is available only to Galaxy admins.
+
+        :type tool_id: str
+        :param tool_id: id of the requested tool
+        """
+        url = "%s/tools/%s/install_dependencies" % (self.gi.url, tool_id)
+        return Client._post(self, payload={}, url=url)
```
Comment (nsoranzo) [inline bioblend/galaxy/tools/__init__.py]: You missed one requested change on this line:
```
s/Client._post(self, /self._post(/
```

Comment (mvdbeek) [inline bioblend/galaxy/tools/__init__.py]: oops, I just amended that.

**Comment by nsoranzo** [comment]
> The API itself is being tested in test/integration/test_resolvers.py, which sets up a conda environment. Can we do this in bioblend?

Is it enough to add `conda_auto_init = True` to the Galaxy config?

**Comment by mvdbeek** [comment]
I think so. Maybe we just test starting from 17.01, where this is default?

**Comment by nsoranzo** [comment]
You can insert:
```bash
# Change Galaxy configuration needed by some tool tests
sed -i -e 's/^#conda_auto_init.*/conda_auto_init = True/' $GALAXY_CONFIG_FILE
```
at line 95 of `run_bioblend_tests.sh` .
I think it should be fine to decorate the forthcoming test with `@test_util.skip_unless_galaxy('release_16.10')` .

File: bioblend/galaxy/tools/__init__.py [hunk 4]
```diff
@@ -66,6 +66,19 @@ def _raw_get_tool(self, in_panel=None, trackster=None):
...
+        Install dependencies for a given tool via a resolver.
+        This works only for Conda currently.
+        This functionality is available since Galaxy release_16.10
+        and is available only to Galaxy admins.
+
+        :type tool_id: str
+        :param tool_id: id of the requested tool
+        """
+        url = "%s/tools/%s/install_dependencies" % (self.gi.url, tool_id)
+        return self._post(self, payload={}, url=url)
```
Comment (nsoranzo) [inline bioblend/galaxy/tools/__init__.py]: Remove the second `self`, this should be:
```python
return self._post(payload={}, url=url)
```

File: tests/TestGalaxyTools.py [hunk 5]
```diff
@@ -126,3 +126,8 @@ def test_run_cat1(self):
         self.assertEqual(len(tool_output["outputs"]), 1)
         # TODO: Wait for results and verify it has 3 lines - 1 2 3, 4 5 6,
         # and 7 8 9.
+
+    @test_util.skip_unless_galaxy('release_16.10')
+    def test_tool_dependency_install(self):
+        installed_dependencies = self.gi.tools.install_dependencies('__SET_METADATA__')
+        self.assertTrue(any([True for d in installed_dependencies if d.get('name', None) == 'samtools' and d.get('dependency_type', None) == 'conda']))
```
Comment (nsoranzo) [inline tests/TestGalaxyTools.py]: You can simplify this as:
```python
self.assertTrue(any(True for d in installed_dependencies if d.get('name') == 'samtools' and d.get('dependency_type') == 'conda'))
```

**Comment by nsoranzo** [comment]
The new test now passes on `release_17.01`, but not on `release_16.10`.

**Comment by mvdbeek** [comment]
Hmm, any way to see if the conda installation succeeded?

**Comment by mvdbeek** [comment]
`requests.packages.urllib3.connectionpool: DEBUG: http://localhost:8080 "POST /api/tools/__SET_METADATA__/install_dependencies?key=21b2ff441f867dc474fb89d310d82b03 HTTP/1.1" 200 None` ... so at least the API itself seems to be OK.

**Comment by nsoranzo** [comment]
Has the output of the API call changed maybe?

**Comment by mvdbeek** [comment]
Slightly, yes, but what we're testing hasn't changed (first call is 16.10):
```
In [3]: gi.tools.install_dependencies('__SET_METADATA__')
Out[3]:
[{u'cacheable': True,
  u'dependency_type': u'conda',
  u'environment_path': None,
  u'exact': True,
  u'model_class': u'CondaDependency',
  u'name': u'samtools',
  u'version': None}]

In [4]: gi.tools.install_dependencies('__SET_METADATA__')
Out[4]:
[{u'cacheable': False,
  u'dependency_type': u'conda',
  u'environment_path': u'/Users/mvandenb/src/galaxy/database/dependencies/_conda/envs/__samtools@_uv_',
  u'exact': True,
  u'model_class': u'MergedCondaDependency',
  u'name': u'samtools',
  u'version': None}]
```

**Comment by nsoranzo** [comment]
Ah, I know why: the change to `run_bioblend_tests.sh` needs to be applied to `.travis.yml` too! Need to integrate these 2 files better.

**Comment by mvdbeek** [comment]
Alright, let's see!

**Comment by nsoranzo** [comment]
Thanks a lot @mvdbeek!

**Pull request merged**